### PR TITLE
Add Builtin.ifdef_<FLAGNAME> as a facility to peek at -D flag that client code is building with

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

File: lib/IRGen/GenBuiltin.cpp [hunk 1]
```diff
@@ -711,6 +711,9 @@ if (Builtin.ID == BuiltinValueKind::id) { \
     return;
   }
 
+  if (Builtin.ID == BuiltinValueKind::Ifdef) {
+    llvm_unreachable("Ifdef not constant folded");
```
Comment (eeckstein) [inline lib/IRGen/GenBuiltin.cpp:715]: nitpick: I'd prefer a error message describing the root cause: "Builtin.ifdef can only be used in @_alwaysEmitIntoClient functions"

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 2]
```diff
@@ -1574,6 +1583,14 @@ void ConstantFolder::initializeWorklist(SILFunction &f) {
         continue;
       }
 
+      // Builtin.ifdef only replaced when not building the stdlib.
+      if (isApplyOfBuiltin(*inst, BuiltinValueKind::Ifdef)) {
+        if (!inst->getModule().isStdlibModule()) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp:1588]: This should depend on `FrontendOptions::ParseStdlib` and not on the module name.
Either you'll add `ParseStdlib` to SILOptions or just use `SILOptions::DisableReplacement` (which is used for the assert configuration).

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 3]
```diff
@@ -1804,13 +1821,32 @@ ConstantFolder::processWorkList() {
...
+    // means we never generate IR for it (when building stdlib). Client code is
+    // then always constant-folding this builtin based on the compilation flags
+    // of the client module.
+    if (isApplyOfBuiltin(*I, BuiltinValueKind::Ifdef)) {
+      if (!I->getModule().isStdlibModule()) {
+        StringRef flagName =
+            cast<BuiltinInst>(I)->getName().str().drop_front(strlen("ifdef_"));
+        const LangOptions &langOpts = I->getModule().getASTContext().LangOpts;
+        bool val = langOpts.isCustomConditionalCompilationFlagSet(flagName);
+        if (constantFoldIfdef(cast<BuiltinInst>(I), val)) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp:1842]: `constantFoldIfdef` never returns an invalid SILValue

File: lib/SILOptimizer/Utils/ConstantFolding.cpp [hunk 4]
```diff
@@ -1152,6 +1152,15 @@ static SILValue constantFoldIsConcrete(BuiltinInst *BI) {
   return inst;
 }
 
+static SILValue constantFoldIfdef(BuiltinInst *BI, bool value) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Utils/ConstantFolding.cpp:1155]: nitpick: maybe call this function `replaceWithIntOne`? Or just inline it

File: stdlib/public/core/Builtin.swift [hunk 5]
```diff
@@ -768,6 +768,12 @@ func _trueAfterDiagnostics() -> Builtin.Int1 {
   return true._value
 }
 
+/// Whether client code is built with -D FOO_BAR, for testing only.
+@_alwaysEmitIntoClient
+private func _ifdefFooBar() -> Bool {
```
Comment (eeckstein) [inline stdlib/public/core/Builtin.swift:773]: Please don't add this function to the stdlib. You can create a test it by compiling a file with `-parse-stdlib` and explicitly `import Swift`.

**Review by eeckstein — CHANGES_REQUESTED** [review]
nice!
Just a few comments.
And: please add a test.

**Comment by kubamracek** [comment]
@eeckstein comments should be addressed now!

**Comment by kubamracek** [comment]
@swift-ci please test

**Review by eeckstein — APPROVED** [review]
lgtm!

**Pull request merged**